torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -34,8 +34,10 @@ void RpcAgent::start() {
 
 void RpcAgent::cleanup() {
   rpcAgentRunning_.store(false);
+  // We must notify the condition variable so it stops waiting in the
+  // retry thread, otherwise this thread cannot be joined.
+  rpcRetryMapCV_.notify_one();
```

Comment:
In general it seems like a good idea to notify without considering `joinable`. Although, I didn't follow how notifying on the condition variable makes the thread joinable? The joinable status of the thread should be orthogonal?

Reply:
Notifying will spuriously wake up the retry thread that is waiting on `rpcRetryMapCV_`: https://github.com/pytorch/pytorch/blob/8ef82fc2c96d59a0c8d65631d77c0ce1c2ed6dea/torch/csrc/distributed/rpc/rpc_agent.cpp#L95 

Since we set `rpcAgentRunning_ ` to False, that sleeping loop will exit from the `retryExpiredRpcs` function here: https://github.com/pytorch/pytorch/blob/8ef82fc2c96d59a0c8d65631d77c0ce1c2ed6dea/torch/csrc/distributed/rpc/rpc_agent.cpp#L88-L89

This ensures the thread has finished execution and can safely be joined.

Reply:
I understand that notifying the CV would ensure the thread has finished execution, but you can call join on a thread even if it has not finished execution. Looking at the example in https://en.cppreference.com/w/cpp/thread/thread/joinable, it seems like joinable can be false if 1) the thread has not started or 2) the thread has already been joined. Could it be possible that the TSAN issue was due to the fact that the thread hadn't started yet and hence joinable was false?

Reply:
Ah I see, I'll do some logging and investigate further